relm4-macros/src/attrs.rs

```diff
@@ -1,31 +1,176 @@
+            leading_colon: Some(Token![::](Span::call_site())),
+            segments: relm4_segments,
+        };
+
+        Attrs {
+            visibility: None,
+
+            relm4_path,
+            relm4_path_set: false,
+        }
     }
 }
 
-impl ToTokens for Attrs {
-    fn to_tokens(&self, out: &mut TokenStream2) {
-        if let Some(vis_pub) = &self.pub_vis {
-            vis_pub.to_tokens(out);
+impl Parse for Attrs {
+    /// Rules for parsing attributes
+    /// 
+    /// 1. It's fine if visibility is used unnamed so `#[widget(pub)]` must be valid but thats the only case
+    /// 2. Widget visibility might be named `#[widget(visibility = pub)]`
+    /// 3. `relm4` argument must be named. Always
+    ///
+    fn parse(input: ParseStream) -> Result<Self> {
+        let mut attrs = Attrs::new();
+        let mut attrs_type = AttributeType::None;
+        let mut dangling_comma;
+
+        let mixed_use_error_message = 
+            "You can't mix named and unnamed arguments while using `relm4_macros::widget`. \n\
+            \n\
+            You can use one of\n\
+            \n\
+            1. `#[relm4_macros::widget()]` to define widget with private visibility\n\
+            2. `#[relm4_macros::widget(pub)]` to define widget with public visibility\n\
+            3. `#[relm4_macros::widget( visibility = pub )]` to define widget with public visibility and potentially other arguments\n\
+            \n\
+            Please use `visibility = pub` to fix this error";
+
+        while !input.is_empty() {            
+            dangling_comma = true;
+
+            if input.peek(Token![pub]) {
+                if matches!(attrs_type, AttributeType::Named) {
+                    return Err(input.error(mixed_use_error_message));
+                }
+                if attrs.visibility.is_some() {
+                    return Err(input.error("You can't assign visibility twice"));
+                }
+                let pub_vis: Visibility = input.parse()?;
+
+                attrs_type = AttributeType::Unnamed{
+                    span: pub_vis.span(),
+                };
+                attrs.visibility = Some(pub_vis);
+                dangling_comma = false;
+            }
+            else if input.peek(Ident) && input.peek2(Token![=]){
+                let ident: Ident = input.parse()?;
+                let eq: Token![=] = input.parse()?;
+
+                if ident == "visibility" {
+                    let pub_vis: Visibility = input.parse()?;
+
+                    if let AttributeType::Unnamed{ span } = attrs_type {
+                        return Err(Error::new(span, mixed_use_error_message));
+                    }
+                    if attrs.visibility.is_some() {
+                        //unwrap's here are safe since all spans are from the same file always
+                        let error_span = ident.span().join(eq.span).unwrap()
+                                            .join(pub_vis.span()).unwrap();
```

Comment:
Does this even run on stable? Maybe the docs are wrong, though: https://docs.rs/proc-macro2/1.0.29/proc_macro2/struct.Span.html#method.join

Reply:
In any case, you don't need all those spans. Just the pub_viz span should be enough to figure out the error.